lib/dao.js

```diff
@@ -624,7 +624,193 @@ DataAccessObject.upsert = function(data, options, cb) {
+              data: update,
+              currentInstance: inst,
+              hookState: ctx.hookState,
+              options: options,
+            };
+            Model.notifyObserversOf('persist', context, function(err) {
+              if (err) return done(err);
+              connector.upsertWithWhere(modelName, where, update, options, done);
+            });
+          }
+          function done(err, data, info) {
+            if (err) return cb(err);
+            var context = {
+              Model: Model,
+              data: data,
+              isNewInstance: info && info.isNewInstance,
+              hookState: ctx.hookState,
+              options: options,
+            };
+            Model.notifyObserversOf('loaded', context, function(err) {
+              if (err) return cb(err);
+              var obj;
+              if (data && !(data instanceof Model)) {
+                inst._initProperties(data, { persisted: true });
+                obj = inst;
+              } else {
+                obj = data;
+              }
 
+              var context = {
+                Model: Model,
+                instance: obj,
+                isNewInstance: info ? info.isNewInstance : undefined,
+                hookState: hookState,
+                options: options,
+              };
+              Model.notifyObserversOf('after save', context, function(err) {
+                cb(err, obj);
+              });
+            });
+          }
+        });
+      } else {
+        var element = Object.keys(where)[0];
+        var filter = { where: where };
+        if (!Object.keys(data).length) {
+          var err = new Error('data object cannot be empty!');
+          err.statusCode = 400;
+          process.nextTick(function() { cb(err); });
+        }
+
+        if (!data.hasOwnProperty(element)) {
+          data[element] = where[element];
+        }
+        var primaryKey = idName(Model);
+        self.find(filter, function(err, instances) {
+          if (err) cb(err);
+          else {
+            var modelsLength = instances.length;
+            if (modelsLength === 0 && data.hasOwnProperty(primaryKey))  {
+              return self.create(data, options, cb);
+            } else if (modelsLength == 0 && data.hasOwnProperty(primaryKey) == false) {
+              var error = new Error('Error message: No matching instance found.Found error while trying ' +
+                'to create a new instance.' + 'Primary key and value ' + 'is not provided in the payload');
+              callback(error);
+            } else if (modelsLength === 1) {
+              var modelInst = instances[0];
+              return modelInst.updateAttributes(data, options, cb);
+            } else {
+              process.nextTick(function() {
+                var error = new Error('There are multiple instances found.' +
+                  'Upsert Operation will not be performed');
+                cb(error);
+              });
```

Comment:
Here is my previous comment which has not been addressed:

`id` can be auto generated `id` in many cases and not provided in `data` object; you should take care of this scenario.

I think you need to remove check `data.hasOwnProperty(pk) === false` and merge this block with the previous one:

```
if (modelsLength === 0) {
  self.create(data, options, cb);
  return cb.promise;
} else if (modelsLength === 1) {
 ...
} else {
 ...
}
```